src/Types/AccountAddress.cs

```diff
@@ -0,0 +1,163 @@
+
+/// <summary>
+/// Represents an account address.
+///
+/// The address of an account is a 32-byte value which uniquely identifies the account.
+/// </summary>
+public readonly struct AccountAddress : IEquatable<AccountAddress>
+{
+    public const uint BytesLength = 32;
+
+    /// <summary>
+    /// A version byte that is prepended to addresses represented as base58 strings.
+    /// </summary>
+    private const byte VersionByte = 1;
+
+    private static readonly Base58CheckEncoder _encoderInstance = new();
+
+    /// <summary>
+    /// Representation of the account address as a byte array (without the version byte prepended).
+    /// </summary>
+    private readonly byte[] _value;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AccountAddress"/> class.
+    /// </summary>
+    /// <param name="addressAsBytes">The address as a sequence of bytes without a version byte prepended.</param>
+    private AccountAddress(byte[] addressAsBytes) => this._value = addressAsBytes;
+
+    /// <summary>
+    /// Get the address represented as a base58 encoded string.
+    /// </summary>
+    public override string ToString() =>
+        // Prepend version byte.
+        _encoderInstance.EncodeData((new byte[] { VersionByte }).Concat(this._value).ToArray());
+
+    /// <summary>
+    /// Get the address as a length-32 byte array without the version byte prepended.
+    /// </summary>
+    public byte[] GetBytes() => (byte[])this._value.Clone();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AccountAddress"/> class.
+    /// </summary>
+    /// <param name="addressAsBase58String">The address represented as a length-50 base58 encoded string.</param>
+    /// <exception cref="ArgumentException">The input is not a length-50 base58 encoded string.</exception>
+    public static AccountAddress From(string addressAsBase58String)
+    {
+        try
+        {
+            var decodedBytes = _encoderInstance.DecodeData(addressAsBase58String).Skip(1).ToArray(); // Remove version byte.
+            return From(decodedBytes);
+        }
+        catch (Exception)
+        {
+            throw new ArgumentException($"'{addressAsBase58String}' is not a length-50 base58 encoded string");
+        }
+    }
+
+    /// <summary>
+    /// Creates an instance from a length-32 byte array representing the address (without the version byte prepended).
+    /// </summary>
+    /// <param name="addressAsBytes">The address as a length-32 byte array, i.e. without the version byte prepended.</param>
+    /// <exception cref="ArgumentException">The input is not a length-32 byte array.</exception>
+    public static AccountAddress From(byte[] addressAsBytes)
+    {
+        if (addressAsBytes.Length != BytesLength)
+        {
+            throw new ArgumentException($"The account address bytes length must be {BytesLength}.");
+        }
+
+        return new AccountAddress((byte[])addressAsBytes.Clone());
```

Comment:
I don’t think we need to clone - arguments are the same as in other clone discussion. If I gave a byte array to the constructor and would expect that byte array to be used as internal representation.

Reply:
Sure, if we actually created the instance from the constructor, I would agree. But the constructor is private.

Instead `From` creates an instance from something (in this case a hex encoded string or a byte array), and abstracts away the implementation details?

Reply:
I changed the method names to `ToBytes`, changed the comment text to indicate that we copy to an array, and instead of cloning, it uses `ToArray`.